modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/TransactionStateResolver.java

```diff
@@ -115,4 +308,64 @@ private void sendAndRetry(CompletableFuture<TxMeta> resFut, ReplicationGroupId r
             }
         });
     }
+
+    /**
+     * Tries to send a request to the given node.
+     *
+     * @param resFut Response future.
+     * @param node Node to send to.
+     * @param request Request.
+     */
+    private void sendAndRetry(CompletableFuture<TransactionMeta> resFut, ClusterNode node, TxStateRequest request) {
+        messagingService.invoke(node, request, RPC_TIMEOUT).thenAccept(resp -> {
+            assert resp instanceof TxStateResponse : "Unsupported response type [type=" + resp.getClass().getSimpleName() + ']';
+
+            TxStateResponse response = (TxStateResponse) resp;
+
+            resFut.complete(response.txStateMeta());
+        });
+    }
+
+    /**
+     * Processes the transaction state requests that are used for coordinator path based write intent resolution. Can't return
+     * {@link org.apache.ignite.internal.tx.TxState#FINISHING}, it waits for actual completion instead.
+     *
+     * @param request Request.
+     * @return Future that should be completed with transaction state meta.
+     */
+    private CompletableFuture<TransactionMeta> processTxStateRequest(TxStateRequest request) {
+        clock.update(request.readTimestamp());
```

Comment:
You should sync it with a chosen commit timestamp because we can generate the timestamp earlier than updating the clock.

Reply:
fixed